File: DataEngineering/08_Data_Governance_and_Quality/projects/python/02_lineage_tracker/main.py
import uuid
import datetime
from openlineage.client import OpenLineageClient, set_producer
from openlineage.client.run import RunEvent, RunState, Run, Job
from openlineage.client.transport.console import ConsoleTransport
from openlineage.client.facet import (
    DocumentationJobFacet, 
    SchemaDatasetFacet, 
    SchemaField, 
    SourceCodeJobFacet
)
from openlineage.client.dataset import Dataset

def main():
    # Set the producer URL indicating the application emitting the events
    set_producer("https://github.com/OpenLineage/OpenLineage/tree/main/integration/python")
    
    # Use a ConsoleTransport to output OpenLineage events to stdout (useful for local testing/demo)
    # In production, use HttpTransport pointing to Marquez or another OpenLineage backend
    client = OpenLineageClient(transport=ConsoleTransport())

    # Define the inputs and outputs for this job run
    input_dataset = Dataset(
        namespace="postgres://oltp-db-prod:5432",
        name="public.raw_transactions",
        facets={
            "schema": SchemaDatasetFacet(
                fields=[
                    SchemaField(name="txn_id", type="INT"),
                    SchemaField(name="amount", type="DECIMAL"),
                    SchemaField(name="user_id", type="INT")
                ]
            )
        }
    )

    output_dataset = Dataset(
        namespace="snowflake://warehouse-prod",
        name="sales_mart.fct_daily_sales",
        facets={
            "schema": SchemaDatasetFacet(
                fields=[
                    SchemaField(name="date", type="DATE"),
                    SchemaField(name="total_amount", type="DECIMAL")
                ]
            )
        }
    )

    job = Job(
        namespace="airflow_production",
        name="daily_sales_aggregation",
        facets={
            "documentation": DocumentationJobFacet(
                description="Aggregates raw transactions into daily sales facts."
            ),
            "sourceCode": SourceCodeJobFacet(
                language="python",
                sourceCode="SELECT date, SUM(amount) FROM raw_transactions GROUP BY date"
            )
        }
    )

    run_id = str(uuid.uuid4())
    run = Run(runId=run_id)

    # Emit a START event
    print("Emitting OpenLineage START event...")
    start_event = RunEvent(
        eventType=RunState.START,
        eventTime=datetime.datetime.now(datetime.timezone.utc).isoformat(),
        run=run,
        job=job,
        inputs=[input_dataset],
        producer="https://github.com/OpenLineage/OpenLineage/tree/main/integration/python"
    )
    client.emit(start_event)

    # Simulate job execution
    print("\nRunning ETL job...")
    
    # Emit a COMPLETE event
    print("\nEmitting OpenLineage COMPLETE event...")
    complete_event = RunEvent(
        eventType=RunState.COMPLETE,
        eventTime=datetime.datetime.now(datetime.timezone.utc).isoformat(),
        run=run,
        job=job,
        outputs=[output_dataset],
        producer="https://github.com/OpenLineage/OpenLineage/tree/main/integration/python"
    )
    client.emit(complete_event)
    print("\nOpenLineage events successfully emitted (printed to console).")

if __name__ == "__main__":
    main()

File: DataEngineering/08_Data_Governance_and_Quality/projects/golang/02_lineage_tracker/main.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/google/uuid"
)

// OpenLineage Core Types
// In a real project, you might generate these from the OpenLineage JSON schemas.

type SchemaField struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

type SchemaDatasetFacet struct {
	Fields []SchemaField `json:"fields"`
}

type DatasetFacets struct {
	Schema *SchemaDatasetFacet `json:"schema,omitempty"`
}

type Dataset struct {
	Namespace string        `json:"namespace"`
	Name      string        `json:"name"`
	Facets    DatasetFacets `json:"facets"`
}

type Job struct {
	Namespace string `json:"namespace"`
	Name      string `json:"name"`
}

type Run struct {
	RunID string `json:"runId"`
}

type RunEvent struct {
	EventType string    `json:"eventType"`
	EventTime string    `json:"eventTime"`
	Run       Run       `json:"run"`
	Job       Job       `json:"job"`
	Inputs    []Dataset `json:"inputs,omitempty"`
	Outputs   []Dataset `json:"outputs,omitempty"`
	Producer  string    `json:"producer"`
}

// emitLineage simulates pushing an OpenLineage event to a backend (like Marquez)
func emitLineage(event RunEvent, endpoint string) error {
	payload, err := json.MarshalIndent(event, "", "  ")
	if err != nil {
		return err
	}

	// For demonstration, we print the payload that would be sent.
	fmt.Printf("Emitting OpenLineage Event to %s:\n%s\n\n", endpoint, string(payload))

	// In a real scenario:
	/*
	req, _ := http.NewRequest("POST", endpoint, bytes.NewBuffer(payload))
	req.Header.Set("Content-Type", "application/json")
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}
	*/
	return nil
}

func main() {
	producerURL := "https://github.com/OpenLineage/OpenLineage/tree/main/integration/golang-custom"
	backendURL := "http://localhost:5000/api/v1/lineage" // Example Marquez HTTP endpoint

	// Define datasets
	inputDB := Dataset{
		Namespace: "postgres://oltp-db-prod:5432",
		Name:      "public.raw_transactions",
		Facets: DatasetFacets{
			Schema: &SchemaDatasetFacet{
				Fields: []SchemaField{
					{Name: "txn_id", Type: "INT"},
					{Name: "amount", Type: "DECIMAL"},
				},
			},
		},
	}

	outputWarehouse := Dataset{
		Namespace: "snowflake://warehouse-prod",
		Name:      "sales_mart.fct_daily_sales",
	}

	// Define Job & Run
	job := Job{
		Namespace: "golang_microservices",
		Name:      "daily_aggregator_worker",
	}
	runID := uuid.New().String()

	// 1. Emit START event
	startEvent := RunEvent{
		EventType: "START",
		EventTime: time.Now().UTC().Format(time.RFC3339),
		Run:       Run{RunID: runID},
		Job:       job,
		Inputs:    []Dataset{inputDB},
		Producer:  producerURL,
	}

	if err := emitLineage(startEvent, backendURL); err != nil {
		log.Fatalf("Failed to emit START event: %v", err)
	}

	// Simulate work...
	fmt.Println("Running data processing logic...")
	time.Sleep(1 * time.Second)

	// 2. Emit COMPLETE event
	completeEvent := RunEvent{
		EventType: "COMPLETE",
		EventTime: time.Now().UTC().Format(time.RFC3339),
		Run:       Run{RunID: runID},
		Job:       job,
		Outputs:   []Dataset{outputWarehouse},
		Producer:  producerURL,
	}

	if err := emitLineage(completeEvent, backendURL); err != nil {
		log.Fatalf("Failed to emit COMPLETE event: %v", err)
	}
}
